Serialize the auto-flow variant of the 'grid' shorthand correctly.

mattwoodrow opened the pull request:
#### d6bea689bd25c1f97669f348eff321757360960d
<pre>
Serialize the auto-flow variant of the &apos;grid&apos; shorthand correctly.
<a href="https://bugs.webkit.org/show_bug.cgi?id=245024">https://bugs.webkit.org/show_bug.cgi?id=245024</a>

Reviewed by Tim Nguyen.

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/parsing/grid-shorthand-serialization-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/parsing/grid-shorthand-serialization.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/parsing/grid-shorthand-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/parsing/grid-shorthand-valid.html:
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::isValueIDIncludingList):
(WebCore::gridAutoFlowContains):
(WebCore::StyleProperties::getGridValue const):
(WebCore::StyleProperties::asTextInternal const):

Canonical link: <a href="https://commits.webkit.org/254974@main">https://commits.webkit.org/254974@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f210c12af872a80ba1f362d38538a05d1e5ac438

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90777 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35355 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21352 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/100076 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/158592 "Reverted pull request changes (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94786 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33855 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28949 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83122 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/96460 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96432 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/26970 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77586 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26776 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81725 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/81473 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69806 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/34943 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15509 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32747 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16490 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3466 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36520 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39434 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38445 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35579 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -984,17 +984,89 @@ String StyleProperties::getGridTemplateValue() const
+
+        return result.toString();
+    }
+
```
Please check that `grid-auto-flow` contains `row`. Otherwise:

```js
document.body.style.cssText = "grid: auto-flow 1px / 2px; grid-auto-flow: inherit";
document.body.style.grid; // "auto-flow 1px / 2px"
```

Loirooriol replied:
Actually, `row` is optional and `dense` is equivalent to `row dense`. So check that it contains either `row` or `dense`, or check that is not a css-wide value, or something like that.

mattwoodrow replied:
`grid` seems like a rather weird shorthand property which doesn't roundtrip properly.

This example (`grid: auto-flow 1px / 2px;`) is almost identical to `Example 27` in the spec, where it says that this value for `grid` also implies `grid-auto-flow: row;`.

To serialize `auto-flow 1px / 2px` would imply that `grid-auto-flow` is `row`, but it's really `inherit`.

As far as I can tell, there is no valid serialisation for `grid` which would correctly display the current style state, which is why I return `""`. Gecko appears to be doing the same thing, and I assume blink too.

Possibly the spec should be more clear on this? It seems like the shorthand isn't really serialisable using normal CSS rules.

mattwoodrow replied:
Apologies, that's exactly what you were suggesting. I'll add tests :)

Loirooriol replied:
Exactly, empty string when the shorthand can't represent all the values of the longhands: https://drafts.csswg.org/cssom/#serializing-css-values

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -984,17 +984,89 @@ String StyleProperties::getGridTemplateValue() const
+    auto autoFlow = getPropertyCSSValue(CSSPropertyGridAutoFlow);
+    if (isValueID(autoColumns, CSSValueAuto)
+        && isValueID(autoRows, CSSValueAuto)
+        && isValueID(autoFlow, CSSValueRow)) {
```
Problem: `grid-auto-flow` can be a list

```js
document.body.style.cssText = "grid: 1px / 2px; grid-auto-flow: row";
document.body.style.grid; // "", expected "1px / 2px"
```

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -984,17 +984,89 @@ String StyleProperties::getGridTemplateValue() const
+    auto autoColumns = getPropertyCSSValue(CSSPropertyGridAutoColumns);
+    auto autoRows = getPropertyCSSValue(CSSPropertyGridAutoRows);
+    auto autoFlow = getPropertyCSSValue(CSSPropertyGridAutoFlow);
+    if (isValueID(autoColumns, CSSValueAuto)
```
Problem: `grid-auto-columns` can be a list

```js
document.body.style.cssText = "grid: 1px / 2px; grid-auto-columns: auto";
document.body.style.grid; // "", expected "1px / 2px"
```

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -984,17 +984,89 @@ String StyleProperties::getGridTemplateValue() const
+    auto autoRows = getPropertyCSSValue(CSSPropertyGridAutoRows);
+    auto autoFlow = getPropertyCSSValue(CSSPropertyGridAutoFlow);
+    if (isValueID(autoColumns, CSSValueAuto)
+        && isValueID(autoRows, CSSValueAuto)
```
Problem: `grid-auto-rows` can be a list

```js
document.body.style.cssText = "grid: 1px / 2px; grid-auto-rows: auto";
document.body.style.grid; // "", expected "1px / 2px"
```

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -984,17 +984,89 @@ String StyleProperties::getGridTemplateValue() const
+    StringBuilder result;
+
+    if (gridAutoFlowContains(autoFlow, CSSValueColumn)) {
+        if (!isValueID(autoRows, CSSValueAuto) || !isValueID(columns.value(), CSSValueNone))
```
Problem: `grid-auto-rows` can be a list

```js
document.body.style.cssText = "grid: 1px / auto-flow 2px; grid-auto-rows: auto";
document.body.style.grid; // "", expected "1px / auto-flow 2px"
```

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -984,17 +984,89 @@ String StyleProperties::getGridTemplateValue() const
+        if (gridAutoFlowContains(autoFlow, CSSValueDense))
+            result.append(" dense");
+
+        if (!isValueID(autoColumns, CSSValueAuto)) {
```
Problem: `grid-auto-columns` can be a list

```js
document.body.style.cssText = "grid: 1px / auto-flow; grid-auto-columns: auto";
document.body.style.grid; // "1px / auto-flow auto", expected "1px / auto-flow"
```

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -984,17 +984,89 @@ String StyleProperties::getGridTemplateValue() const
+        return result.toString();
+    }
+
+    if (!isValueID(autoColumns, CSSValueAuto) || !isValueID(rows.value(), CSSValueNone))
```
Problem: `grid-auto-columns` can be a list

```js
document.body.style.cssText = "grid: auto-flow 1px / 2px; grid-auto-columns: auto";
document.body.style.grid; // "", expected "auto-flow 1px / 2px"
```

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -984,17 +984,89 @@ String StyleProperties::getGridTemplateValue() const
+    if (gridAutoFlowContains(autoFlow, CSSValueDense))
+        result.append(" dense");
+
+    if (!isValueID(autoRows, CSSValueAuto)) {
```
Problem: `grid-auto-rows` can be a list

```js
document.body.style.cssText = "grid: auto-flow dense / 2px; grid-auto-rows: auto";
document.body.style.grid; // auto-flow dense auto / 2px", expected "auto-flow dense / 2px"
```

Loirooriol reviewed (changes requested):
Please handle list values. And would be good to add test coverage, since Blink is failing these cases too.

Loirooriol commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/parsing/grid-shorthand-valid.html:
```diff
@@ -51,6 +52,41 @@
+test_valid_combined_shorthand_value("grid: 1px / auto-flow 2px; grid-auto-rows: auto", "grid", "1px / auto-flow 2px");
+test_valid_combined_shorthand_value("grid: 1px / auto-flow; grid-auto-columns: auto", "grid", "1px / auto-flow");
+test_valid_combined_shorthand_value("grid: auto-flow 1px / 2px; grid-auto-columns: auto", "grid", "auto-flow 1px / 2px");
+test_valid_combined_shorthand_value("grid: auto-flow dense / 2px; grid-auto-rows: auto", "grid", "auto-flow dense / 2px");
```
Nit: since these use multiple declarations, it's more about serializing the `grid` shorthand rather than parsing as valid, so may be better to move them into css/css-grid/parsing/grid-shorthand-serialization.html

Could also be good to include some cases that you handled properly but Blink didn't: https://bugs.chromium.org/p/chromium/issues/detail?id=1353822#c7

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -984,17 +990,92 @@ String StyleProperties::getGridTemplateValue() const
+        return String();
+
+    if (!autoColumns || !autoRows)
+        return String();
```
Nit: this could be moved just after defining `autoColumns` and `autoRows`? And I guess it wouldn't hurt to check  `!autoFlow` too.

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -984,17 +990,81 @@ String StyleProperties::getGridTemplateValue() const
+            return String();
+
+        if (isPropertyImplicit(CSSPropertyGridTemplateRows))
+            return String();
```
I'm not sure if I get this. https://searchfox.org/wubkat/rev/cf0cfbaf1ad7728a145c06e3539170870bbb913f/Source/WebCore/css/CSSProperty.h#58 says
```cpp
uint16_t m_implicit : 1; // Whether or not the property was set implicitly as the result of a shorthand.
```

Why wouldn't we want to serialize if it has been set from a shorthand? In fact, it serializes after setting the `grid` shorthand, so the bit is not set I guess? Not sure if the description is wrong.

If this condition can be removed, then it's one less search operation in the vector of properties.

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -984,17 +990,81 @@ String StyleProperties::getGridTemplateValue() const
+    auto rows = getPropertyCSSValue(CSSPropertyGridTemplateRows);
+    auto columns = getPropertyCSSValue(CSSPropertyGridTemplateColumns);
+    if (!rows || !columns)
+        return String();
```
This is between the declaration of `autoColumns`/`autoRows`/`autoFlow` and some checks for `autoColumns`/`autoRows`/`autoFlow`.

Seems to me it would make more sense to move it below. `getGridTemplateValue()` already checks whether the properties are present anyways.

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -77,8 +77,14 @@ static bool isNormalValue(const RefPtr<CSSValue>& value)
+static bool isValueID(const RefPtr<CSSValue>&, CSSValueID);
 static bool isValueID(const Ref<CSSValue>& value, CSSValueID id)
 {
+    if (is<CSSValueList>(value)) {
+        if (downcast<CSSValueList>(value.get()).size() != 1)
+            return false;
+        return isValueID(downcast<CSSValueList>(value.get()).item(0), id);
+    }
```
Modifying `isValueID` to allow lists can make it inconsistent with other checks like `isNormalValue` or `isNoneValue`.

Maybe add `isValueIDIncludingList` or something.

Loirooriol commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/parsing/grid-shorthand-serialization.html:
```diff
@@ -0,0 +1,112 @@
+<!DOCTYPE html>
```
This test should keep the existing checks?

mattwoodrow replied:
Ah, looks like my local wpt was out of date and only had two subtests here.

I had to change a few of the expected results, which hopefully are correct.

The first subtest was setting `grid: auto-flow auto / 100px 100px` and expecting `auto-flow auto / 100px 100px`, which I think fails the shortest serialisation principle, and `auto` is unnecessary. I changed it to accept `auto-flow / 100px 100px` or `none / 100px 100px` which seem equivalent, and I can't find a way to determine which should take precedence.

The third subtest was setting `grid: auto-flow / 10px; grid-template-rows: 20px` and expecting no serialisation for `grid`. I think since it's setting `grid-auto-flow: rows` (the default), we can still serialise to `20px / 10px`.

Loirooriol replied:
Sounds good, this should make the test pass in Firefox.

Loirooriol commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/parsing/grid-shorthand-serialization.html:
```diff
@@ -0,0 +1,137 @@
+
+    const should = (value === "") ? "; shouldn't" : "; should";
+    test(function(){
+      try {
```
Is the try...finally necessary?

Loirooriol commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/parsing/grid-shorthand-serialization.html:
```diff
@@ -0,0 +1,137 @@
+</head>
+<body>
+<script>
+  function test_shorthand_roundtrip(cssText, property, value, longhands, declarations) {
```
The `longhands` already contains some shorthand like `grid-template` in some cases, then maybe it should be renamed and `property` and `value` should be merged into that? So

```js
  test_shorthand_roundtrip('grid: auto-flow auto / 100px 100px',
                            'grid', ['auto-flow / 100px 100px', 'none / 100px 100px'],
                            {
    'grid-template-areas': 'none'
  });
```

could just be

```js
  test_shorthand_roundtrip('grid: auto-flow auto / 100px 100px', {
    'grid': ['auto-flow / 100px 100px', 'none / 100px 100px'],
    'grid-template-areas': 'none',
  });
```

Loirooriol commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/parsing/grid-shorthand-serialization.html:
```diff
@@ -0,0 +1,137 @@
+      let cssText = div.style.cssText;
+      declarations.forEach(decl => {
+        try {
+          assert_true(cssText.indexOf(decl) !== -1, `cssText ('${cssText}') must contain '${decl}'`);
```
This should be inside a `test()`.

Loirooriol commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/parsing/grid-shorthand-serialization.html:
```diff
@@ -0,0 +1,137 @@
+    'grid-auto-flow: row;',
+    'grid-auto-columns: auto;',
+    'grid-auto-rows: auto;',
+  ]);
```
This was already wrong, but if `grid` serializes, then `cssText` should just be `grid: 20px / 10px`?

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -1729,6 +1802,11 @@ StringBuilder StyleProperties::asTextInternal() const
+            case CSSPropertyGridTemplateRows:
+            case CSSPropertyGridTemplateColumns:
+            case CSSPropertyGridTemplateAreas:
+                shorthandPropertyID = CSSPropertyGridTemplate;
```
It's fine to leave it for a follow-up, but this should try to serialize with the `grid` shorthand if possible. You can copy the `serializeBorderShorthand` hack to handle the multiple shorthands.

Loirooriol reviewed (approved):
Informal LGTM

mattwoodrow committed:
Serialize the auto-flow variant of the 'grid' shorthand correctly.
https://bugs.webkit.org/show_bug.cgi?id=245024

Reviewed by Tim Nguyen.

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/parsing/grid-shorthand-serialization-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/parsing/grid-shorthand-serialization.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/parsing/grid-shorthand-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/parsing/grid-shorthand-valid.html:
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::isValueIDIncludingList):
(WebCore::gridAutoFlowContains):
(WebCore::StyleProperties::getGridValue const):
(WebCore::StyleProperties::asTextInternal const):

Canonical link: https://commits.webkit.org/254974@main
